codec-http2/src/test/java/io/netty/handler/codec/http2/Http2FrameCodecTest.java

```diff
@@ -666,6 +668,47 @@ public void upgradeEventNoRefCntError() throws Exception {
         assertEquals(1, upgradeEvent.refCnt());
     }
 
+    @Test
+    public void upgradeWithoutFlowControlling() throws Exception {
+        channel.pipeline().addAfter(http2HandlerCtx.name(), null, new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelRead(final ChannelHandlerContext ctx, Object msg) throws Exception {
+                if (msg instanceof Http2DataFrame) {
+                    // Simulate consuming the frame and update the flow-controller.
+                    Http2DataFrame data = (Http2DataFrame) msg;
+                    ctx.writeAndFlush(new DefaultHttp2WindowUpdateFrame(data.initialFlowControlledBytes())
+                            .stream(data.stream())).addListener(new ChannelFutureListener() {
+                        @Override
+                        public void operationComplete(ChannelFuture future) throws Exception {
+                            Throwable cause = future.cause();
+                            if (cause != null) {
+                                ctx.fireExceptionCaught(cause);
+                            }
+                        }
+                    });
+                }
+                ReferenceCountUtil.release(msg);
+            }
+        });
+
+        frameListener.onHeadersRead(http2HandlerCtx, Http2CodecUtil.HTTP_UPGRADE_STREAM_ID, request, 31, false);
+
+        // Using reflect as the constructor is package-private and the class is final.
+        Constructor<UpgradeEvent> constructor =
+                UpgradeEvent.class.getDeclaredConstructor(CharSequence.class, FullHttpRequest.class);
+
+        // Check if we could make it accessible which may fail on java9.
+        Assume.assumeTrue(ReflectionUtil.trySetAccessible(constructor) == null);
+
+        String longString = new String(new char[70000]).replace("\0", "*");
+        DefaultFullHttpRequest request =
+            new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/", bb(longString));
+
+        HttpServerUpgradeHandler.UpgradeEvent upgradeEvent = constructor.newInstance(
+            "HTTP/2", request);
+        channel.pipeline().fireUserEventTriggered(upgradeEvent);
```

Comment:
this test doesn't drive `Http2ConnectionHandler` through the necessary series of events to prepare for the upgrade. `onHttpServerUpgrade` is never called (and there is some stuff that must occur before that). Consider creating a unit test which uses `Http2ServerUpgradeCodec` (or simulate the events necessary) so `Http2ConnectionHandler`'s stat is properly updated.

Reply:
@Scottmitch I don't understand this, and unfortunately don't have time to tackle it.  Instead of letting this PR languish, how about I revert this commit, go back to the previous implementation, which worked but didn't cover all cases, and I'll file a ticket to fix it later.  Otherwise, I'll try to return to this when I have time (probably sometime next year).

Reply:
> I don't understand this

The test is not following the same steps that `Http2ServerUpgradeCodec` would, and therefore state is not properly initialized. Your code changes which invoke `designateCleartextUpgradeConnection` are not called.

> how about I revert this commit, go back to the previous implementation

The approach seems correct but if you don't want to fix the unit test reverting is fine as long as https://github.com/netty/netty/pull/7400#discussion_r152634982 is addressed.

> I've also borked my local version, so tests hang every time I run them

Can you clarify what "local version" means here, and which tests hang? Netty tests, finagle tests, other tests?

Reply:
I have a clone of the netty repository, and the netty tests hang.